library/signed/src/Operations.qs

```diff
@@ -0,0 +1,361 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import Std.Arrays.Tail, Std.Arrays.Most, Std.Arrays.Enumerated;
+import Utils.AndLadder;
+import Unstable.Arithmetic.RippleCarryTTKIncByLE;
+import Std.Diagnostics.Fact;
+import Comparison.CompareGTI;
+
+/// # Summary
+/// Square signed integer `xs` and store
+/// the result in `result`, which must be zero initially.
+///
+/// # Input
+/// ## xs
+/// 𝑛-bit integer to square
+/// ## result
+/// 2𝑛-bit result, must be in state |0⟩
+/// initially.
+///
+/// # Remarks
+/// The implementation relies on `SquareI`.
+operation SquareSI(xs : Qubit[], result : Qubit[]) : Unit is Adj + Ctl {
+    body (...) {
+        Controlled SquareSI([], (xs, result));
+    }
+    controlled (controls, ...) {
+        let n = Length(xs);
+        use signx = Qubit();
+        use signy = Qubit();
+
+        within {
+            CNOT(Tail(xs), signx);
+            Controlled Invert2sSI([signx], xs);
+        } apply {
+            Controlled SquareI(controls, (xs, result));
+        }
+    }
+}
+
+/// # Summary
+/// Computes the square of the integer `xs` into `result`,
+/// which must be zero initially.
+///
+/// # Input
+/// ## xs
+/// 𝑛-bit number to square
+/// ## result
+/// 2𝑛-bit result, must be in state |0⟩ initially.
+///
+/// # Remarks
+/// Uses a standard shift-and-add approach to compute the square. Saves
+/// 𝑛-1 qubits compared to the straight-forward solution which first
+/// copies out `xs` before applying a regular multiplier and then undoing
+/// the copy operation.
+operation SquareI(xs : Qubit[], result : Qubit[]) : Unit {
+    body (...) {
+        Controlled SquareI([], (xs, result));
+    }
+    controlled (controls, ...) {
+        let n = Length(xs);
+
+
+        let numControls = Length(controls);
+        if numControls == 0 {
+            use aux = Qubit();
+            for (idx, ctl) in Enumerated(xs) {
+                within {
+                    CNOT(ctl, aux);
+                } apply {
+                    Controlled AddI([aux], (xs, (result[idx..idx + n])));
+                }
+            }
+        } elif numControls == 1 {
+            use aux = Qubit();
+            for (idx, ctl) in Enumerated(xs) {
+                within {
+                    AND(controls[0], ctl, aux);
+                } apply {
+                    Controlled AddI([aux], (xs, (result[idx..idx + n])));
+                }
+            }
+        } else {
+            use helper = Qubit[numControls];
```

Comment:
nit: `helper` could be renamed to `aux` to be consistent with the other conditional blocks.